python/kserve/kserve/protocol/rest/openai/openai_model.py

```diff
@@ -300,6 +302,161 @@ def mapper(completion: Completion) -> ChatCompletionChunk:
+        """Postprocess a completion. Only called when response is not being streamed (i.e. stream=false)"""
+        pass
+
+    def postprocess_completion_chunk(self, completion: Completion):
+        """Postprocess a completion chunk. Only called when response is being streamed (i.e. stream=true)
+        This method will be called once for each chunk that is streamed back to the user.
+        """
+        pass
+
+    def preprocess_chat_completion_request(self, request: ChatCompletionRequest):
+        """Preprocess a chat completion request."""
+        pass
+
+    def postprocess_chat_completion(self, chat_completion: ChatCompletion):
+        """Postprocess a chat completion. Only called when response is not being streamed (i.e. stream=false)"""
+        pass
+
+    def postprocess_chat_completion_chunk(
+        self, chat_completion_chunk: ChatCompletionChunk
+    ):
+        """Postprocess a chat completion chunk. Only called when response is being streamed (i.e. stream=true)
+        This method will be called once for each chunk that is streamed back to the user.
+        """
+        pass
+
+    def _handle_completion_chunk(self, raw_chunk: str):
+        # Skip empty lines
+        if len(raw_chunk) == 0:
+            return None
+        # All chunks are prefixed with "data: "
+        data = raw_chunk[6:]
+        if data == "[DONE]":
+            return None
+        completion_chunk = Completion.model_validate_json(data)
+        self.postprocess_completion_chunk(completion_chunk)
+        return completion_chunk
+
+    def _handle_chat_completion_chunk(self, raw_chunk: str):
+        # Skip empty lines
+        if len(raw_chunk) == 0:
+            return None
+        # All chunks are prefixed with "data: "
+        if len(raw_chunk) == 0:
+            return None
+        data = raw_chunk[6:]
+        if data == "[DONE]":
+            return None
+        chat_completion_chunk = ChatCompletionChunk.model_validate_json(data)
+        self.postprocess_chat_completion_chunk(chat_completion_chunk)
+        return chat_completion_chunk
+
+    async def create_completion(
+        self, request: CompletionRequest
+    ) -> Union[Completion, AsyncIterator[Completion]]:
+        self.preprocess_completion_request(request)
+        if request.params.stream:
+            async with self._http_client.stream(
+                "POST",
+                self._completions_endpoint,
+                content=request.params.model_dump_json(),
+                headers={"Content-type": "application/json"},
+            ) as r:
+                it = AsyncMappingIterator(
+                    iterator=r.aiter_lines(), mapper=self._handle_completion_chunk
+                )
+                return it
+        else:
+            response = await self._http_client.post(
+                self._completions_endpoint,
+                content=request.params.model_dump_json(),
+                headers={"Content-type": "application/json"},
+            )
```

Comment:
I initially assumed we would instantiate an OpenAI client here and use it to send the requests. I guess this is more future proof. Do you see any potential down sides of this? Or any advantages to use the client lib?

Reply:
I think the main reason we are sending requests with an HTTP client is so we don't need to add another dependency. We have generated Pydantic types from the OpenAI swagger schema so we don't need the client for validation.

Reply:
We used to have that dependency anyway, if openai is enabled. I discovered that, that behavior changed at https://github.com/kserve/kserve/pull/3589/files#diff-20e43250cf05a057d2394b9ed1a4462b95403b7945198976c7e288913902b58a

Sounds all good.